PORTAL/jobs/_utils.py

```diff
@@ -508,42 +593,54 @@ def _from_values(cls, elapsed, units):
         return self
 
     @classmethod
-    def _convert_to_seconds(cls, elapsed, units):
+    def _convert_to_seconds(
+            cls,
+            elapsed,
+            units: Optional[str]
+    ) -> "ElapsedTimeWithUnits":
         if units == 's':
             return elapsed
         return (elapsed / cls._resolve_units(units)).normalize()
 
     @classmethod
-    def _convert_from_seconds(cls, elapsed, units):
+    def _convert_from_seconds(
+            cls,
+            elapsed,
+            units: str
+    ) -> "ElapsedTimeWithUnits":
         if units == 's':
             return elapsed
         return (elapsed * cls._resolve_units(units)).normalize()
 
     @classmethod
-    def _resolve_units(cls, units):
-        try:
-            return cls._UNITS[units]
-        except KeyError:
+    def _resolve_units(cls, units: Optional[str]) -> int:
+        if units not in cls._UNITS:
             raise ValueError(f'unsupported units {units!r}')
+        return cls._UNITS[units]
 
     @classmethod
-    def _validate_elapsed(cls, elapsed):
+    def _validate_elapsed(cls, elapsed) -> None:
         if elapsed < 0:
             raise ValueError(f'expected non-negative value, got {elapsed}')
 
     @classmethod
-    def _check_for_abnormal(cls, elapsed, units):
+    def _check_for_abnormal(cls, elapsed, units: str) -> None:
         if elapsed < 1 or elapsed >= 1000:
             logger.warn('abnormal elapsed value {elapsed} {units}, consider normalizing')
 
-    def __new__(cls, elapsed, units='s'):
+    def __new__(
+            cls,
+            elapsed,
+            units: str = 's'
+    ) -> "ElapsedTimeWithUnits":
         self = super().__new__(cls)
         self._elapsed = decimal.Decimal(elapsed)
         self._units = units
         return self
 
     def __init__(self, *args, **kwargs):
         self._elapsed = self._elapsed.normalize()
+        self._units: Optional[str] = None
```

Comment:
This would override what was set in `__new__()`.

Reply:
Good catch.  I think it just needs a declaration, not a value set.

Reply:
Would you be okay with keeping attribute declarations up at the class level?  It's easier to miss otherwise.

Reply:
Sure.